sound/soc/sof/intel/hda.c

```diff
@@ -410,9 +410,16 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 			pdata->tplg_filename =
 				hda_mach->sof_tplg_filename;
 
-			/* firmware: pick the first in machine list */
+			/*
+			 * firmware: pick the first in machine list,
+			 * or use nocodec
+			 */
 			mach = pdata->desc->machines;
-			pdata->fw_filename = mach->sof_fw_filename;
+			if (mach)
```

Comment:
@plbossart Im not sure an empty struct is equivalent to NULL. Maybe check for ARRAY_SIZE() is a better option?

Reply:
yes, that was my fear when I wrote the code, but then we also test for the NULL case when we reach the last empty structure in the table. That's why I'd want someone to test on CFL w/ HDA.

Reply:
@plbossart one way to make sure that this is null is not to set the "machines" field until we have entries in the cfl table.
But in my multi-client series, I fixed this problem by adding a default_fw_filename field to desc because the firmware is loaded long before the acpi match is done by the audio client. But thats for another day.

Reply:
right, and for now I am stuck due to this other issue:
https://github.com/thesofproject/linux/issues/1417

whoever invented signed firmware should buy free beer for life to all developers.

Reply:
if (mach->id[0]) is what we used for acpi scans, so fixed the code to use the same thing